src/api.jl

```diff
@@ -12,114 +12,259 @@ import .OptimizeElbo
 """
-function infer(
-      dir::AbstractString, run::AbstractString, camcol::AbstractString,
-      field::AbstractString, outdir::AbstractString,
-      partnum::Int, parts::Int)
-
-    # get images
-    images = SkyImages.load_sdss_blob(dir, run, camcol, field)
-
-    # load catalog and convert to Array of `CatalogEntry`s.
-    cat_df = SDSS.load_catalog_df(dir, run, camcol, field)
-    cat_entries_all = SkyImages.convert_catalog_to_celeste(cat_df, images)
-
-    # Filter low-flux objects.
-    flux_cols = [ symbol("psfflux_$b") for b in band_letters ]
-    max_fluxes =
-      Float64[ maximum([ cat_df[row, flux_col] for flux_col in flux_cols ])
-               for row in 1:size(cat_df, 1) ];
-    good_rows = max_fluxes .> 2;
-    println("Keeping $(sum(good_rows) / length(good_rows)) of the objects.")
-    cat_entries = cat_entries_all[good_rows]
-
-    # limit to just the part of the catalog specified.
-    partsize = length(cat_entries) / parts
-    minidx = round(Int, partsize * (partnum - 1)) + 1
-    maxidx = round(Int, partsize * partnum)
-
-    objids = [ entry.objid for entry in cat_entries[minidx:maxidx] ]
+function infer(ra_range::Tuple{Float64, Float64},
+               dec_range::Tuple{Float64, Float64},
+               fieldids::Vector{Tuple{Int, Int, Int}},
+               frame_dirs::Vector;
+               fpm_dirs=frame_dirs,
+               psfield_dirs=frame_dirs,
+               photofield_dirs=frame_dirs,
+               photoobj_dirs=frame_dirs)
+
+    # the code below assumes there is at least one field.
+    if length(fieldids) == 0
+        return Dict{Int, Dict}()
+    end
+
+    # Read in all photoobj catalogs.
+    rawcatalogs = Array(Dict, length(fieldids))
+    for i in eachindex(fieldids)
+        run, camcol, field = fieldids[i]
+        dir = photoobj_dirs[i]
+        fname = @sprintf "%s/photoObj-%06d-%d-%04d.fits" dir run camcol field
+        info("field $i: reading $fname")
+        rawcatalogs[i] = SDSSIO.read_photoobj(fname)
+    end
+
+    for i in eachindex(fieldids)
+        info(string("field ", i, ": ", length(rawcatalogs[i]["objid"]),
+                    " entries"))
+    end
+
+    # Limit each catalog to primary objects
+    for cat in rawcatalogs
+        mask = cat["mode"] .== 0x01
+        for key in keys(cat)
+            cat[key] = cat[key][mask]
+        end
+    end
+
+    for i in eachindex(fieldids)
+        info(string("field ", i, ": ", length(rawcatalogs[i]["objid"]),
+                    " primary entries"))
+    end
+
+    # Merge all catalogs together (there should be no duplicate objects,
+    # because for each object there should only be one "primary" occurance.)
+    rawcatalog = deepcopy(rawcatalogs[1])
+    for i=2:length(rawcatalogs)
+        for key in keys(rawcatalog)
+            append!(rawcatalog[key], rawcatalogs[i][key])
+        end
+    end
+
+    # Limit catalog to ra, dec range.
+    mask = ((rawcatalog["ra"] .> ra_range[1]) &
```

Comment:
If there is a celestial object near the edge of an [RA, DEC] rectangle that contributes photons to an image but is not included in this rectangle, I think this would exclude it from inference, right?  That will lead to problems similar to what I described in https://github.com/jeff-regier/Celeste.jl/issues/130.